envsource.go

```diff
@@ -0,0 +1,465 @@
+		// because the PkgPath is also null for them.
+		// ref: https://github.com/golang/go/issues/21122
+		if field.PkgPath != "" { //field is unexported
+			continue
+		}
+
+		values, err := e.analyzeValue(field.Type, append(currentPath, field.Name))
+
+		if err != nil {
+			return []*envValue{}, err
+		}
+
+		res = append(res, values...)
+	}
+
+	return res, nil
+}
+
+func (e *envSource) analyzeValue(valType reflect.Type, fieldPath path) ([]*envValue, error) {
+	var (
+		res []*envValue
+		err error
+	)
+	switch valType.Kind() {
+	case reflect.Array, reflect.Slice, reflect.Map:
+		res, err = e.analyzeIndexedType(valType, fieldPath)
+	case reflect.Ptr:
+		res, err = e.analyzeValue(valType.Elem(), fieldPath)
+	case reflect.Struct:
+		res, err = e.analyzeStruct(valType, fieldPath)
+	case reflect.Invalid, reflect.Chan, reflect.Func, reflect.Interface, reflect.UnsafePointer:
+		//Skip these fields, don't throw...
+		//TODO : keep track of the fields ignored by the library to be able to list them.
+		//err = fmt.Errorf("type %s is not supported by EnvSource. fieldPath : %v", valType.Name(), fieldPath)
+	default:
+		res = e.loadValue(fieldPath)
+	}
+
+	return res, err
+}
+
+func (e *envSource) analyzeIndexedType(valType reflect.Type, fieldPath path) ([]*envValue, error) {
+	var (
+		res []*envValue
+	)
+
+	prefix := e.envVarFromPath(fieldPath)
+	vars := e.envVarsWithPrefix(prefix)
+	nextKeys := unique(e.nextLevelKeys(prefix, vars))
+
+	for _, varName := range nextKeys {
+		key := e.keyFromEnvVar(varName, prefix)
+
+		// If we're on an Int based key, we need to be able to convert
+		// detected key to an int
+		if valType.Kind() == reflect.Array ||
+			valType.Kind() == reflect.Slice {
+			index, err := strconv.Atoi(key)
+
+			if err != nil {
+				return res, fmt.Errorf(
+					key,
+					varName,
+				)
+			}
+
+			if valType.Kind() == reflect.Array &&
+				index >= valType.Len() {
+				return res, fmt.Errorf(
+					"Detected key (%s) from variable %s is >= to array length %d",
+					key,
+					varName,
+					valType.Len(),
+				)
+			}
+		}
+
+		valPath := append(fieldPath, key)
+		keyValues, err := e.analyzeValue(valType.Elem(), valPath)
+
+		if err != nil {
+			return res, err
+		}
+
+		res = append(res, keyValues...)
+	}
+
+	return res, nil
+}
+
+func (e *envSource) loadValue(fieldPath path) []*envValue {
+	variableName := e.envVarFromPath(fieldPath)
+
+	value, ok := os.LookupEnv(variableName)
+
+	if !ok {
+		return []*envValue{}
+	}
+
+	return []*envValue{&envValue{value, fieldPath.clone()}}
+}
+
+func (e *envSource) assignValues(configVal reflect.Value, envValues []*envValue, currentPath []string) error {
+
+	if len(currentPath) > 0 {
+		envValues = filterEnvVarWithPrefix(envValues, currentPath)
+	}
+
+	if configVal.Kind() == reflect.Ptr {
+		if configVal.IsNil() {
+			configVal.Set(reflect.New((configVal.Type().Elem())))
+		}
+		err := e.assignValues(configVal.Elem(), envValues, []string{})
+		return err
```

Comment:
```go
return e.assignValues(configVal.Elem(), envValues, nil)
```

Reply:
Wouldn't `if len(currentPath) > 0 {` fail if I pass `nil` as a parameter to this recursion?
Is it so usual to pass around `nil` in golang? It seems like a recipe to create `nil` related error, no?

Reply:
a slice `nil` have a `len` equals to `0`